<a href="https://colab.research.google.com/github/SofiMont1/Anal-tica-de-negocios/blob/main/NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

*Modelo de ApreAprobación - Naive Bayes**
0. Se procede con la carga de librerías de trabajo

In [ ]:
import numpy as np #Funciones Matemáticas
import pandas as pd #Comunicación con Excel
from sklearn.naive_bayes import GaussianNB #Función Gaussiana
from sklearn.metrics import confusion_matrix #La métrica de medida


1. Se cargan los datos

In [ ]:
nxl='/content/SolicitantesCrédito(USD).xlsx'
XDB=pd.read_excel(nxl,sheet_name=2)
XDB.dropna() #Eliminar datos faltantes, celdas vacias
print(XDB) #Mostrar la base de datos

#Seleccionar las variables de interes
XDB=XDB[['Edad','Ingresos','Egresos','Cuota (COP)','PreApr']]

#Datos de Entrada - np.array() - Convertir a numeros
XD=np.array(XDB[['Edad','Ingresos','Egresos','Cuota (COP)']])

#Datos de Salida - Valor deseado - Que quiero que aprenda el modelo
yd=np.array(XDB['PreApr'])


      Unnamed: 0  Edad    Sexo Estado civil  Hijos  Perscargo  Nivel Renta  \
0              1    38      NR       Casado      1          0            0   
1              2    51      NR     Separado      6          1            1   
2              3    37   Mujer       Casado      2          0            0   
3              4    29      NR     Separado      1          0            0   
4              5    42   Mujer       Casado      3          4            4   
...          ...   ...     ...          ...    ...        ...          ...   
5837        8372    48  Hombre     Separado      0          0            0   
5838        8374    31  Hombre      Soltero      2          0            0   
5839        8375    38  Hombre     Separado      0          0            0   
5840        8376    43   Mujer       Casado      1          2            2   
5841        8377    37   Mujer       Casado      1          2            2   

      Estrato Nivel de estudios  Tipo de contrato  ...  Plazo  

2. Implementamos Modelo

In [ ]:
from functools import cmp_to_key
mnb=GaussianNB() #LLamamos el modelo
mnb.fit(XD,yd)
ydp=mnb.predict(XD) #muetsra los datos de salida - pero que el obtuvo

#Digame las medidas de las variables para cada categoria - PreApr/PreNeg
u=mnb.theta_ #Se estiman las medidas
sigma=np.sqrt(mnb.var_) #Se estiman las varianzas

#Las variaciones de las variables
LI=u-sigma
LS=u+sigma

print("La media de las variables por categoría:")
print(u)
print("El limite inferior de la variación de las variables por categoria es:")
print(LI)
print("El limite superior de la variación de las variables por categoria es:")
print (LS)

#Métrica de Evaluación - Como se comporto el modelo identificando la PreApr-PreNeg
cm=confusion_matrix(yd,ydp)

La media de las variables por categoría:
[[  40.98411626  431.42680644  401.54995644   16.88529983]
 [  41.10752688 1029.98947251  900.19312144   43.3269839 ]]
El limite inferior de la variación de las variables por categoria es:
[[ 28.77108358 105.62889147  58.77709557 -12.88412769]
 [ 28.85646196 557.55604444 374.75974307 -21.65436704]]
El limite superior de la variación de las variables por categoria es:
[[  53.19714893  757.22472141  744.32281731   46.65472735]
 [  53.35859181 1502.42290058 1425.62649981  108.30833484]]


4. Se procede con la evaluación del modelo

In [ ]:
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdaderos Positivos

Exactitud=(VP+VN)/len(XD)
print("La exactitud del modelo es: ",Exactitud) #Comportamiento general

TE=(FN+FP)/len(XD)
print("La tasa de error es: ",TE) #Porcentaje equivociones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad es:" ,Sensibilidad) #Como se comporto frente a los positivos

Especificidad=VN/(VN+FN)
print("Especificidad es: ",Especificidad) #Como se comportaron frente a los negativos

Precision=VP/(VP+FN)
print("La precisión es: ",Precision) #Como se comportaron frente a los positivos

PredNeg=VN/(VN+FP)
print("La probabilidad de Negativo es: ",PredNeg) #Como se comportaron frente a los negativos


La exactitud del modelo es:  0.7281752824375214
La tasa de error es:  0.2718247175624786
La sensibilidad es: 0.8102539530426449
Especificidad es:  0.6825565912117177
La precisión es:  0.5865417967395075
La probabilidad de Negativo es:  0.8661710037174721


5. Que pasa con una sola persona

In [ ]:
Xun=np.array([[38, 1356, 1686, 18]]) #Esta la persona del Excel
ydun=mnb.predict(Xun)
print(ydun)


[1]


Análisis de Resultados
De los resultados se puede obervar que los preaprobados en general presentan edades(), ingresos (), egresos (), y cuotas () mayores a los aprobados. Con respecto a las metricas se puede observar que el modelo funcionó mejor clasificando los positivos, como dice el indicador de sensibilidad el cual alcanzó el 81% de certeza. De manera general se puede observar que el modelo logró con exactitud 70% en promedio, y la tasa de error cercana al 30% en promedio. De manera general esta exactitud lo convierte en un buen modelo. Con respecto al solicitante de crédito tal y tal este pertenece a al categoría de aprobado.